src/middlewared/middlewared/plugins/cluster_linux/ctdb_private.py

```diff
@@ -0,0 +1,241 @@
+
+        # make sure that the cluster shared volume is mounted
+        if not await self.middleware.call('service.started', 'glusterd'):
+            verrors.add(
+                f'{schema_name}.glusterd',
+                'The "glusterd" service is not started.',
+            )
+
+        try:
+            mounted = pathlib.Path(SHARED_VOL).is_mount()
+        except Exception:
+            mounted = False
+
+        if not mounted:
+            verrors.add(
+                f'{schema_name}.{SHARED_VOL}',
+                f'"{SHARED_VOL}" is not mounted'
+            )
+
+        verrors.check()
+
+        # get the current ips in the cluster
+        cur_ips = await self.middleware.call('ctdb.private.ips.query')
+        cur_ips.extend((await self.middleware.call('ctdb.public.ips.query')))
+
+        if not delete:
+            # make sure private ip doesn't already exist in the cluster
+            if pri_ip in cur_ips:
+                verrors.add(
+                    f'{schema_name}.{pri_ip}',
+                    f'"{pri_ip}" already in the cluster.'
+                )
+        else:
+            # make sure private ip exists before deleting from the cluster
+            if pri_ip not in cur_ips:
+                verrors.add(
+                    f'{schema_name}.{pri_ip}',
+                    f'"{pri_ip}" does not exist in the cluster.'
+                )
+
+        verrors.check()
+
+    @private
+    def update_file(self, pri_ip, verrors, delete=False):
+        """
+        Update the ctdb cluster private IP file.
+        """
+
+        # ctdb documentation is _VERY_ explicit in
+        # how the private IP file is modified
+
+        # the documentation clearly states that before
+        # adding a private peer, the cluster must be
+        # healthy. This requires running a command that
+        # is expecting the ctdb daemon to be started.
+        # If this is the first private peer being added
+        # then the ctdb daemon isn't going to be started
+        # which means we can't check if the cluster is
+        # healthy. So we do the following:
+        #   1. if the ctdb shared volume private ip file exists
+        #       then assume that this isn't the first peer
+        #       being added to the cluster and check the ctdb
+        #       daemon for the cluster health.
+        #   2. elif the ctdb shared volume private ip doesnt
+        #       exist then assume this is the first peer being
+        #       added to the cluster and skip the cluster health
+        #       check.
+        ctdb_file = pathlib.Path(CTDBConfig.GM_PRI_IP_FILE.value)
+        etc_file = pathlib.Path(CTDBConfig.ETC_PRI_IP_FILE.value)
+
+        if ctdb_file.exists():
+            if self.middleware.call_sync('service.started', 'ctdb'):
+                if not self.middleware.call_sync('ctdb.general.healthy'):
+                    self.logger.error('ctdb cluster is not healthy')
```

Comment:
might be worth adding a note on what it implies: `ctdb cluster is not healthy, not updating private ip files`